### ISSUE TITLE
Shortcode’s don’t always have access to doc.templateContent

### ISSUE THREAD

**Issue by kleinfreund** [11ty/eleventy#479]
For a while now, I’m struggling to write a properly functioning excerpt shortcode. The following worked for a while:

```js
eleventyConfig.addShortcode('excerpt', post => extractExcerpt(post));
```

```js
function extractExcerpt(doc) {
  if (!doc.hasOwnProperty('templateContent')) {
    console.warn('❌ Failed to extract excerpt: Document has no property `templateContent`.');
    return;
  }

  const excerptSeparator = '<!--more-->';
  const content = doc.templateContent;

  if (content.includes(excerptSeparator)) {
    return content.substring(0, content.indexOf(excerptSeparator)).trim();
  }

  const pCloseTag = '</p>';
  if (content.includes(pCloseTag)) {
    return content.substring(0, content.indexOf(pCloseTag) + pCloseTag.length);
  }

  return content;
}
```

... but that stopped working a while ago. I was too lazy to investigate it closely and hoped it would magically fix itself.

Anyway, I’m getting an `Tried to use templateContent too early` error for my blog index page:

```
Error writing templates: (more in DEBUG output)
> Having trouble writing template: _site/index.html (TemplateWriterWriteError)
> Tried to use templateContent too early (./src/index.md page 0) (TemplateContentPrematureUseError):
    TemplateContentPrematureUseError: Tried to use templateContent too early (./src/index.md page 0)
        at Object.get templateContent [as templateContent] (/home/phil/dev/repos/kleinfreund.de/node_modules/@11ty/eleventy/src/Template.js:452:21)
        at Template.renderPageEntry (/home/phil/dev/repos/kleinfreund.de/node_modules/@11ty/eleventy/src/Template.js:530:57)
        at Template.writeMapEntry (/home/phil/dev/repos/kleinfreund.de/node_modules/@11ty/eleventy/src/Template.js:546:32)
        at TemplateWriter._writeTemplate (/home/phil/dev/repos/kleinfreund.de/node_modules/@11ty/eleventy/src/TemplateWriter.js:133:6)
        at TemplateWriter.write (/home/phil/dev/repos/kleinfreund.de/node_modules/@11ty/eleventy/src/TemplateWriter.js:164:24)
```

**TL;DR**: I want to be able to *access the templace content in Markdown stage* in order to find a `<!--more-->` string and generate an excerpt with it. This currently only works on the HTML stage where this comment string is no longer there.

**Comment by kleinfreund** [comment]
For features like shortcode, we need to clearly document *when* they are executed. Since they are called from templates, I expect shortcodes to run from documents before they were transformed into HTML. In other words, not having access to `templateContent` is a bug, I think.

**Comment by Ryuno-Ki** [comment]
Maybe related to #406

**Comment by zachleat** [comment]
@kleinfreund can you include an example of how you’re calling the shortcode?

**Comment by kleinfreund** [comment]
```liquid
---
layout: _includes/layouts/base.liquid
css:
- components.pagination.css
pagination:
  data: collections.posts
  size: 5
  reverse: true
---
<div class="post-list">
  {%- for post in pagination.items -%}
  <article class="post">

    <!-- some meta data stuff -->

    {% excerpt post %}

  </article>
  {%- endfor -%}
</div>

<!-- pagination -->
```

**Comment by kleinfreund** [comment]
In fact, this breaks my build because the main `index.html` is no longer generated. 😱

**Comment by zachleat** [comment]
Hmmmmmm… Eleventy 0.8.x builds a dependency graph based on pagination and collection tags so this is certainly a bug.

A ha, I see what’s happening. Hmm, looks like the issue happens when the pagination keys are in a layout. If you move that content into a main template (temporarily) it will work for now.

**Comment by kleinfreund** [comment]
Indeed, moving the layout content I posted in https://github.com/11ty/eleventy/issues/479#issuecomment-480616822 to my `index.md` and renaming it to `index.liquid` works around the issue.

**Comment by zachleat** [comment]
The fix for this issue will be included with 0.8.3.
